README.md
# @davidfengau/date-picker

React Date Picker initially built for use in BGL SF360.

## Installation and usage

```
yarn add @davidfengau/date-picker
```

Then use it in your app:

```js
import React from 'react'
import moment from 'moment'
import DatePicker from '@davidfengau/date-picker'

class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      startDate: moment()
    }
    this.handleChange = this.handleChange.bind(this)
  }
  
  handleChange(date) {
    this.setState({
      startDate: date
    })
  }
    
  render() {
    return (
      <DatePicker
        onChange={this.handleChange}
        selected={this.state.Date}
        className="form-control"
      />
    )
  }
}
```

Use Date Range Picker:
```js
import { DateRangePicker } from '@davidfengau/date-picker'
```

## Props

Common props you may want to specify include:

- `onChange` - subscribe to date change
- `selected`  - control the current value
- `tabIndex` - specify the tab index
- `className` - apply a className to the component
- `minDate` - Min Date allowed to select
- `maxDate` - Max Date allowed to select
- `placeholderText` - string
- `allowSameDay` - boolean
- `showYearDropdown` - boolean

Extra props for date range picker:

- `onChange` - subscribe to date change
- `selectsStart` - boolean
- `selectsEnd` - boolean
- `startDate` - moment
- `endDate` - moment
- `startDatePlaceholder` - string
- `endDatePlaceholder` - string
- `showClearAll` - boolean

## Development
```js
yarn install
yarn run build
```

## Publish
```js
# For public npm
npm publish --access public

# For BGL only
npm publish --registry http://repo.bglcorp.com.au:8081/content/repositories/bgl.npm/
```


CHANGELOG.md
# @davidfengau/date-picker

## 1.0.9, 1.0.8, 1.0.7
- hot fix: selected type error

## 1.0.6
- remove Fund from date range picker and replace with FundFYStartDate, FundFYEndDate, FundFormedDate

## 1.0.5
- hot fix: selected type error

## 1.0.4
- hot fix: in case of no selected date provide

## 1.0.2
- hot fix: specify granularity in moment.js comparison 

## 1.0.1
- hot fix: export DateRangePicker in Typings

## 1.0.0
- release








src/lib/Fund.js
import moment from 'moment'
import { calcFinancialYear, currentFinancialYear } from './Utils'

class Fund {
  fyStartDate
  fyEndDate
  formedDate
  currentFY

  constructor(fyStartDate, fyEndDate, formedDate) {
    this.fyStartDate = fyStartDate
    this.fyEndDate = fyEndDate
    this.formedDate = formedDate
    this.currentFY = currentFinancialYear
  }

  isEmpty() {
    return this.fyStartDate === undefined && this.fyEndDate === undefined && this.formedDate === undefined
  }

  startDate() {
    return this.fyStartDate || this.formedDate || moment(`${this.currentFY - 1}-07-01`)
  }

  endDate() {
    return this.fyEndDate || moment(`${this.currentFY}-06-30`)
  }

  financialYear() {
    return this.fyEndDate ? calcFinancialYear(this.fyEndDate) : calcFinancialYear(moment())
  }
}

export default Fund


src/lib/index.d.ts
import * as moment from 'moment'
import * as React from 'react'

declare function transformDate(value: string): moment.Moment | null

declare function calcFinancialYear(date: moment.Moment): number

declare const currentFinancialYear: number

export interface IDatePickerProps {
  onChange: (date: moment.Moment | undefined) => void,
  selected?: moment.Moment | undefined,
  tabIndex?: number | undefined,
  className?: string | undefined,
  minDate?: moment.Moment | undefined,
  maxDate?: moment.Moment | undefined,
  placeholderText?: string | undefined,
  allowSameDay?: boolean | undefined,
  showYearDropdown?: boolean | undefined
}

export interface IDatePickerStates {
  value: string
}

declare class DatePicker extends React.Component<IDatePickerProps, IDatePickerStates> {
  onChange(date: moment.Moment | null): void
  onBlur(event: React.FocusEvent<HTMLInputElement>): void
  onKeyDown(event: React.KeyboardEvent<HTMLDivElement>): void
  onChangeRaw(event: React.FocusEvent<HTMLInputElement>): void
}

export default DatePicker

export interface IDateRangePickerProps {
  onChange: (dateRange: { startDate: moment.Moment | undefined, endDate: moment.Moment | undefined }) => void,
  selectsStart?: boolean | undefined,
  selectsEnd?: boolean | undefined,
  startDate?: moment.Moment | undefined,
  endDate?: moment.Moment | undefined,
  startDatePlaceholder?: string | undefined,
  endDatePlaceholder?: string | undefined,
  showClearAll?: boolean | undefined,
  fundFYStartDate?: moment.Moment | undefined,
  fundFYEndDate?: moment.Moment | undefined,
  fundFormedDate?: moment.Moment | undefined
}

export declare class DateRangePicker extends React.Component<IDateRangePickerProps> {
  onChange(startDate?: moment.Moment | undefined, endDate?: moment.Moment | undefined): void
  onChangeStartDate(startDate: moment.Moment): void
  onChangeEndDate(endDate: moment.Moment): void
  onClearDate(): void
  setDatesByFY(startDate: moment.Moment, endDate: moment.Moment): void
}


src/lib/DatePicker.js
import moment from 'moment'
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import MomentPropTypes from 'react-moment-proptypes'
import ReactDatePicker from 'react-datepicker'
import { E_KEY, ENTER_KEY, F_KEY, H_KEY, HOME_BACKWARD_KEY, HOME_FORWARD_KEY, K_KEY, L_KEY, M_KEY, MINUS_KEY,
  MINUS_KEY_2, MINUS_KEY_3, PAGE_DOWN_KEY, PAGE_UP_KEY, PLUS_KEY, PLUS_KEY_2, PLUS_KEY_3, R_KEY, T_KEY, W_KEY,
  Y_KEY } from './KeysCode'
import { currentFinancialYear, transformDate } from './Utils'

class DatePicker extends Component {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    selected: MomentPropTypes.momentObj,
    tabIndex: PropTypes.number,
    className: PropTypes.string,
    minDate: MomentPropTypes.momentObj,
    maxDate: MomentPropTypes.momentObj,
    placeholderText: PropTypes.string,
    allowSameDay: PropTypes.bool,
    showYearDropdown: PropTypes.bool
  }

  constructor(props) {
    super(props)
    this.state = {
      value: props.selected ? props.selected.format('DD/MM/YYYY') : ''
    }
  }

  componentDidUpdate(prevProps) {
    const preDate = prevProps.selected
    const thisDate = this.props.selected
    if ((!moment.isMoment(preDate && thisDate) && moment.isMoment(thisDate || preDate))
      || (preDate && thisDate && !preDate.isSame(thisDate, 'date'))) {
      this.setState({
        value: thisDate ? thisDate.format('DD/MM/YYYY') : ''
      })
    }
  }

  isInRange = (date) => {
    return date === null || date === undefined || (
      (!this.props.minDate || date.isSameOrAfter(this.props.minDate, 'date'))
      && (!this.props.maxDate || date.isSameOrBefore(this.props.maxDate, 'date'))
    )
  }

  onChange = (date) => {
    const checkedDate = this.isInRange(date) ? date : this.props.selected
    if (checkedDate === null || this.props.selected === null || this.props.selected === undefined
      || (checkedDate && !checkedDate.isSame(this.props.selected, 'date'))) {
      this.props.onChange(date)
    }
    this.setState({ value: checkedDate ? checkedDate.format('DD/MM/YYYY') : '' })
  }

  onBlur = (event) => {
    if (event && event.target) {
      this.onChange(transformDate(event.target.value))
    }
  }

  onKeyDown = (event) => {
    switch (event.keyCode) {
      case ENTER_KEY:
        // @ts-ignore
        this.onBlur(event)
        break
      case T_KEY:
        this.onChange(moment())
        break
      case PLUS_KEY:
      case PLUS_KEY_2:
      case PLUS_KEY_3:
        this.onChange(moment(this.props.selected).add(1, 'days'))
        break
      case MINUS_KEY:
      case MINUS_KEY_2:
      case MINUS_KEY_3:
        this.onChange(moment(this.props.selected).subtract(1, 'days'))
        break
      case PAGE_DOWN_KEY:
        this.onChange(moment(this.props.selected).add(1, 'months'))
        break
      case PAGE_UP_KEY:
        this.onChange(moment(this.props.selected).subtract(1, 'months'))
        break
      case HOME_FORWARD_KEY:
        this.onChange(moment(this.props.selected).add(1, 'years'))
        break
      case HOME_BACKWARD_KEY:
        this.onChange(moment(this.props.selected).subtract(1, 'years'))
        break
      case F_KEY:
        this.onChange(moment(`${currentFinancialYear - 1}-07-01`))
        break
      case L_KEY:
      case E_KEY:
        this.onChange(moment(`${currentFinancialYear}-06-30`))
        break
      case W_KEY:
        this.onChange(moment().startOf('week'))
        break
      case K_KEY:
        this.onChange(moment().endOf('week'))
        break
      case M_KEY:
        this.onChange(moment().startOf('month'))
        break
      case H_KEY:
        this.onChange(moment().endOf('month'))
        break
      case Y_KEY:
        this.onChange(moment().startOf('year'))
        break
      case R_KEY:
        this.onChange(moment().endOf('year'))
        break
      default:
    }
  }

  onChangeRaw = (event) => {
    this.setState({
      value: event.target.value.replace(/[^0-9/]+/g, '')
    })
  }

  render() {
    return (
      <ReactDatePicker
        {...this.props}
        ref={(ref) => { this.datePicker = ref }}
        dateFormat="DD/MM/YYYY"
        value={this.state.value}
        onBlur={this.onBlur}
        onKeyDown={this.onKeyDown}
        onChange={this.onChange}
        onChangeRaw={this.onChangeRaw}
      />
    )
  }
}

export default DatePicker


src/lib/DateRangePicker.js
import moment from 'moment'
import React from 'react'
import PropTypes from 'prop-types'
import MomentPropTypes from 'react-moment-proptypes'
import Fund from './Fund'
import DatePicker from './DatePicker'

class DateRangePicker extends React.Component {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    startDate: MomentPropTypes.momentObj,
    endDate: MomentPropTypes.momentObj,
    startDatePlaceholder: PropTypes.string,
    endDatePlaceholder: PropTypes.string,
    showClearAll: PropTypes.bool,
    fundFYStartDate: MomentPropTypes.momentObj,
    fundFYEndDate: MomentPropTypes.momentObj,
    fundFormedDate: MomentPropTypes.momentObj
  }

  onChange = (startDate, endDate) => {
    this.props.onChange({ startDate, endDate })
  }

  onChangeStartDate = (startDate) => {
    this.onChange(startDate, this.props.endDate)
  }

  onChangeEndDate = (endDate) => {
    this.onChange(this.props.startDate, endDate)
  }

  onClearDate = () => {
    this.onChange()
    this.datePickerStart.datePicker.handleCalendarClickOutside()
    this.datePickerEnd.datePicker.handleCalendarClickOutside()
  }

  setDatesByFY = (startDate, endDate) => {
    this.onChange(startDate, endDate)
    this.datePickerStart.datePicker.handleCalendarClickOutside()
    this.datePickerEnd.datePicker.handleCalendarClickOutside()
  }

  datePeriodOptions = () => {
    const fund = new Fund(this.props.fundFYStartDate, this.props.fundFYEndDate, this.props.fundFormedDate)
    const fy = fund.financialYear()
    return (
      <div className="datePeriodOptions">
        <div key="currentFY" onClick={() => this.setDatesByFY(fund.startDate(), fund.endDate())}>
          {fund.isEmpty() ? 'Current Financial Year' : 'Current Fund FY'}
        </div>
        <div key="nextFY" onClick={() => this.setDatesByFY(moment(`${fy}-07-01`), moment(`${fy + 1}-06-30`))}>
          {fund.isEmpty() ? 'Next Financial Year' : 'Next Fund FY'}
        </div>
        {this.props.showClearAll ? <div className="clearAll" onClick={this.onClearDate}>Clear Dates</div> : null}
      </div>
    )
  }

  render() {
    return [
      <DatePicker
        ref={(ref) => { this.datePickerStart = ref }}
        key="startDate"
        className="form-control start"
        selected={this.props.startDate}
        selectsStart
        startDate={this.props.startDate}
        endDate={this.props.endDate}
        onChange={this.onChangeStartDate}
        maxDate={this.props.endDate}
        showYearDropdown
        placeholderText={this.props.startDatePlaceholder || 'Start Date'}
      >
        {this.datePeriodOptions()}
      </DatePicker>,
      <span key="separator" className="separator">-</span>,
      <DatePicker
        ref={(ref) => { this.datePickerEnd = ref }}
        key="endDate"
        className="form-control end"
        selected={this.props.endDate}
        selectsEnd
        startDate={this.props.startDate}
        endDate={this.props.endDate}
        onChange={this.onChangeEndDate}
        minDate={this.props.startDate}
        allowSameDay
        showYearDropdown
        placeholderText={this.props.endDatePlaceholder || 'End Date'}
      >
        {this.datePeriodOptions()}
      </DatePicker>]
  }
}

export default DateRangePicker
